Source/WebCore/html/ImageDataArray.cpp

```diff
@@ -29,6 +29,165 @@
 #include "config.h"
 #include "ImageDataArray.h"
 
+#include <JavaScriptCore/Float16Array.h>
+#include <JavaScriptCore/Uint8ClampedArray.h>
+
+// Needed for `downcast` below.
+SPECIALIZE_TYPE_TRAITS_BEGIN(JSC::Uint8ClampedArray)
+    static bool isType(const JSC::ArrayBufferView& arrayBufferView) { return arrayBufferView.getType() == JSC::TypeUint8Clamped; }
+SPECIALIZE_TYPE_TRAITS_END()
+SPECIALIZE_TYPE_TRAITS_BEGIN(JSC::Float16Array)
+    static bool isType(const JSC::ArrayBufferView& arrayBufferView) { return arrayBufferView.getType() == JSC::TypeFloat16; }
+SPECIALIZE_TYPE_TRAITS_END()
+
 namespace WebCore {
 
+template <typename F>
+static auto visitArrayBufferView(JSC::ArrayBufferView& bufferView, F&& f)
+{
+    // Always try Uint8ClampedArray first, as it should be the most frequent.
+    if (auto* array = dynamicDowncast<JSC::Uint8ClampedArray>(bufferView))
+        return std::forward<F>(f)(*array);
+    if (auto* array = dynamicDowncast<JSC::Float16Array>(bufferView))
+        return std::forward<F>(f)(*array);
+    RELEASE_ASSERT_NOT_REACHED("Unexpected ArrayBufferView type");
+}
+
+ImageDataArray::ImageDataArray(Ref<JSC::ArrayBufferView>&& arrayBufferView)
+    : m_arrayBufferView(WTFMove(arrayBufferView))
+{
+    ASSERT(isSupported(m_arrayBufferView.get()));
+}
+
+ImageDataArray::ImageDataArray(Ref<JSC::Uint8ClampedArray>&& data)
+    : ImageDataArray(Ref<JSC::ArrayBufferView>(WTFMove(data)))
+{ }
+
+ImageDataArray::ImageDataArray(Ref<JSC::Float16Array>&& data)
+    : ImageDataArray(Ref<JSC::ArrayBufferView>(WTFMove(data)))
+{ }
```

Comment:
The issue is that a random `ArrayBufferView` could be other than uint8 or float16, which we don't support.
These two constructors cannot fail, and they ensure that `ImageDataArray` is always valid by construction.

An alternative would be to have something like `static std::optional<ImageDataArray> tryCreate(Ref<JSC::ArrayBufferView>&&)`, but I think it would make life unnecessarily more difficult for callers that already have a `Uint8ClampedArray` or `Float16Array` at hand.

So I'll keep these for now, but we can always revisit if something else becomes more appropriate as development continues...